### PR TITLE
Enable basic Kept validation in NativeAOT running linker tests

### PR DESCRIPTION
**Pull request by vitek-karas**
This implements basic Type and Method Kept attribute validation.
It adds a `KeptBy` property on all `KeptAttribute` which uses the same `ProducedBy` enum to specify exceptions for each target product. (Eventually after the source merge we should rename `ProducedBy` enum to just `Tool` or `Platform`, but it's not important and I didn't want to do it here as it would be a giant diff).

The validation is done by going over all `ConstructedEETypeNode` and `IMethodBodyNode` nodes in the final graph and remembers that list. Then it compares that list to the expected kept members as per the attributes in tests. There are some tweaks:
* Filters out all compiler added types/members
* Doesn't track generic instantiations - only remember definition
* If a method body is kept, then it behaves as if the type is also kept even though there's no `ConstructedEETypeNode` - this is technically wrong (since NativeAOT can remove the type if for example only static methods are used from it), but it would require major changes to the linker tests (since in IL this is a necessity). If we ever need precise validation of this, we would probably add a new attribute to check just for this.

I also copied all of the "other assemblies" kept validation code from ResultChecker (were it lives in linker) to AssemblyChecker where it will need to be to actually work (and where it should probably be anyway). That code is not used yet.

Left lot of TODO/commented out code in the AssemblyChecker - lots of other validation to enable eventually.

Fixed/adapted all of the tests which were already ported to NativeAOT to pass with this new validation.
Removed a test for unresolved members, since it actually fails NativeAOT compilation, so it doesn't test anything really.

One tiny product change:
Display names now consistently include all parent types when writing out nested type names. ILLink already does this and NativeAOT was a bit inconsistent. Since the display names are used only in diagnostics, this brings the diagnostics experience closer between the two tools. The added benefit is that we can use display names to compare members between Cecil and Internal.TypeSystem more precisely.

### CODE REVIEW

File: src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs [hunk 1]
```diff
@@ -36,7 +36,8 @@ public static string GetDisplayName(this MethodDesc method)
 
             if (method.IsConstructor)
             {
-                sb.Append(method.OwningType.GetDisplayNameWithoutNamespace());
+                DefType defType = method.OwningType.GetTypeDefinition() as DefType;
+                sb.Append(defType?.Name ?? method.Name);
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs:40]: I'll just assume the changes here match what IL Linker is doing. But for the `??` part I want to point out that we would only take that codepath for array constructors (so e.g. `new int[10, 20, 30];`) and whether that part is intentional.

If so, I'd suggest we change the if above to `if (method.IsConstructor && method.OwningType is DefType defType)`. Array constructors will fall down to the last `if` in the cascade.

File: src/coreclr/tools/aot/Mono.Linker.Tests.Cases.Expectations/Assertions/KeptAttribute.cs [hunk 2]
```diff
@@ -8,5 +8,11 @@ namespace Mono.Linker.Tests.Cases.Expectations.Assertions
 	[AttributeUsage (AttributeTargets.All, Inherited = false)]
 	public class KeptAttribute : BaseExpectedLinkedBehaviorAttribute
 	{
+		/// <summary>
+		/// By default the target should be kept by all platforms
+		/// This property can override that by setting only the platforms
+		/// which are expected to keep the target.
+		/// </summary>
+		public ProducedBy KeptBy { get; set; } = ProducedBy.TrimmerAnalyzerAndNativeAot;
```
Comment (marek-safar) [inline src/coreclr/tools/aot/Mono.Linker.Tests.Cases.Expectations/Assertions/KeptAttribute.cs:16]: nit: kind of weird name to repeat the name of the attribute on the property.

Comment (vitek-karas) [reply src/coreclr/tools/aot/Mono.Linker.Tests.Cases.Expectations/Assertions/KeptAttribute.cs:16]: So just `[Kept (By = Tool.Trimmer)]`? (After we rename the enum...)
I can change that, sounds good.

**Review by sbomer — APPROVED** [review]
Very nice, thanks! LGTM on first pass. I can do another more detailed review tomorrow if you like, but the fact that it already caught a bunch of differences between linker/nativeaot is great.

File: src/coreclr/tools/aot/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs [hunk 3]
```diff
@@ -70,6 +109,55 @@ protected AssemblyDefinition ResolveOriginalsAssembly (string assemblyName)
...
+				//	throw new NotImplementedException ($"Unknown assembly assertion of type {assemblyAttr.AttributeType}");
+			}
+
+			//if (trimModeIsCopy) {
+			//	foreach (string assemblyName in Directory.GetFiles (Directory.GetParent (outputDirectory).ToString (), "input")) {
+			//		var fileInfo = new FileInfo (assemblyName);
+			//		if (fileInfo.Extension == ".dll" && !actionAssemblies.Contains (assemblyName))
+			//			VerifyCopyAssemblyIsKeptUnmodified (outputDirectory, assemblyName + (assemblyName == "test" ? ".exe" : ".dll"));
+			//	}
+			//}
```
Comment (tlakollo) [inline src/coreclr/tools/aot/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs:150]: Nit: I think I like the practice of commenting big pieces of code using a #if directive (see [DescriptorMarker.cs](https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DescriptorMarker.cs#L144-L150)), will likely avoid the SA1027 too.

File: src/coreclr/tools/aot/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs [hunk 4]
```diff
@@ -45,18 +45,57 @@ public ResultChecker ()
...
+		{
+			InitializeResolvers (testResult);
 
 			try {
-				var original = ResolveOriginalsAssembly (trimmedResult.ExpectationsAssemblyPath.FileNameWithoutExtension);
-				AdditionalChecking (trimmedResult, original);
+				var original = ResolveOriginalsAssembly (testResult.ExpectationsAssemblyPath.FileNameWithoutExtension);
+
+				if (!HasAttribute (original, nameof (NoLinkedOutputAttribute))) {
+					// TODO Validate presense of the main assembly - if it makes sense (reflection only somehow)
```
Comment (tlakollo) [inline src/coreclr/tools/aot/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs:72]:
```suggestion
					// TODO Validate presence of the main assembly - if it makes sense (reflection only somehow)
```

File: src/coreclr/tools/aot/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs (hunk 3 above)
Comment (vitek-karas) [reply src/coreclr/tools/aot/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs]: I'm... dumb... thanks for this suggestion.

File: src/coreclr/tools/aot/Mono.Linker.Tests/Extensions/CecilExtensions.cs [hunk 5]
```diff
@@ -310,6 +311,7 @@ public static StringBuilder GetDisplayNameWithoutNamespace (this TypeReference t
 					break;
 				}
 
+				type = type.GetElementType ();
```
[the remark this answers is not in the document]
Comment (vitek-karas) [reply src/coreclr/tools/aot/Mono.Linker.Tests/Extensions/CecilExtensions.cs:314]: In Cecil this method is something like "Get me the type without the decorators". So for arrays it returns the element type, but it also works for things like pointer types (where it returns the non-pointer version), by-ref types and so on. And on all other "normal" types it just returns `this`.

The case I ran into was a by-ref of a nested type. Before this change we would write it out as simple `NestedType&`, after the change it prints out `ParentType.NestedType&`.

**Pull request merged**